### ISSUE TITLE
Transmit input to dreadnaut (nauty) via file

### ISSUE THREAD

**Issue by TJenrich** [gap-packages/grape#17]
In the function `SetAutGroupCanonicalLabelingNauty` as defined in grape.g of version 4.8.3 of GRAPE, the input for the external program dreadnaut is written into a string (named fdre). 
In the "alternative" function `SetAutGroupCanonicalLabelingBliss`, the input for the external program is written into a file. 
While (for the current GAP version) the use of a stream is considerably faster than using a file, it can stress the memory limits, lead to crashing GAP, start massive swapping, even up to the point that turning off the computer seems the best solution. 
To me, it became a problem because for a certain application the data to feed dreadnaut is about 700 MiB.  Of course, the problems as well as the system capabilities can vary a lot. 
Therefore I propose to let it to the user whether a string or a file is to use. 
I successfully tested that idea by copying the original grape package into the user's subdirectory .gap/pkg and modifying grape.g in four places: 
Just before the definition of `SetAutGroupCanonicalLabelingNauty`, I inserted 
```gap
 nauty_input_via_file:=false;
```
Within the body of that function, the three changed pieces are
```gap
 if nauty_input_via_file then
   fdre:=Filename(GRAPE_nautytmpdir,"ftmp0");
   fdre_stream:=OutputTextFile(fdre,false);
  else
   fdre:="";
   fdre_stream:=OutputTextString(fdre,false);
  fi;
```

```gap
  if nauty_input_via_file then
   GRAPE_Exec(GRAPE_DREADNAUT_EXE, [], InputTextFile(fdre),  OutputTextUser());
  else
   GRAPE_Exec(GRAPE_DREADNAUT_EXE, [], InputTextString(fdre), OutputTextUser());
  fi;
```

```gap
  if nauty_input_via_file then
   RemoveFile(fdre);
  fi;
```
While the first two fragments should replace (easy to find) original instructions, the latter is to insert just before the end of the body of the function.

Of course, the integration into the official GRAPE package would make it simpler than such a workaround, in particular if one wants to publish a GAP/GRAPE based solution.

**Comment by lhsoicher** [comment]
@TJenrich Thank you for pointing out this inconsistency in GRAPE.

This was introduced in commit f0f4b59b42d8bccc07b637a4a72c0947daed3053 which formed
part of PR #12 submitted by @fingolfin . Perhaps @fingolfin can comment on the different treatments 
of both input and output to nauty/dreadnaut and to bliss. It looks like a string is used for nauty/dreadnaut 
input, but a file is used for bliss input, and a file is used for nauty/dreadnaut output, but a string is used for
bliss output. I would appreciate @fingolfin commenting on this.
In particular, were all these changes essential for "to make it work in the Windows virtual machines provided by AppVeyor"?

Thank you @TJenrich also for your suggested changes.
I don't think I ever needed to input as much as 700MB of data to a run of nauty/dreadnaut!
One solution for you would be to try using bliss. I have had very good results using bliss,
in particular for testing block designs for isomorphism (via the the DESIGN package), and did not
notice a problem using a file for input.

**Comment by fingolfin** [comment]
As to the specific reasoning for changes I made in PR #12: it's been 1.5 years, I can't say I remember, but my guess would be that I simply restricted myself to minimal, "obviously right" changes; and the original code passed input to nauty via stdin (by piping a file into stdin), and used a command line option (I guess?) to specify the output file; while for bliss, it did the reverse: specify the input file via a command line option, and read the output from stdout (by piping it into a file and then reading that file). And I simply did the trivial changes of modifying those two piped streams to be in-memory instead of using files. I don't know why the original code (which I didn't write) makes the choices it does...

I understand that this can be a problem if you have huge graphs as input; but of course if you have instead many *many* small graphs, then avoiding temporary files can be better. So perhaps an option to choose this method (for expert users) would be a good idea. One may want to consider doing for both input and outputs of both tools, though?

Note that the overall best approach to deal with these issues, at least on POSIX systems, would likely be to use `InputOutputLocalProcess`, i.e., directly pass the data to the child program, instead of feeding it first into a GAP string or a file; that should avoid both memory issues as well as overhead from creating and deleting many temporary files. However, I am not sure whether this works on Windows right now (as in: whether GAP implements it there -- based on what I read in our test suite, I guess not... Although, now that I am reminded about and poke at it, I *do* wonder: I made some changes to that code last year which should in principle make it work on Windows... Gotta find somebody on Windows who can test it...

Alas, switching to `InputOutputLocalProcess` of course will require additional work (and testing), which somebody has to invest.

**Comment by TJenrich** [comment]
Thanks for the quick and informative reactions. Meanwhile I followed the suggestion of Leonard to try to choose bliss instead of nauty for my purposes (namely, let GRAPE calculate the CliqueNumber of graphs in the size class of the $Fi_{23}$ graph (on 31671 vertices)), though I am familiar with nauty but not with bliss. The results were disappointing: Because of the less efficient format, input to bliss is twice the size of input to dreadnaut (for the same graph). In my test it were about 1.1 GiB from an srg(28671,3150). But the real problem is that bliss takes much more memory than dreadnaut. For the mentioned example, 2.8 GiB were the last size that I could see in the process manager before bliss crashed less than 30 seconds after its start (on my system with two cores, 4 GiB Ram, Lubuntu 20.04), even after modifying grape.g such that bliss is fed with the (half-sized) undirected graph, and even when I saved the file fdre before GRAPE removed it and later fed it to bliss from the command line (without running GAP or other big programs in the background). So, I stick to nauty (and my proposal).
